### PR TITLE
[db engine] Add support for Elasticsearch

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Add support for Elasticsearch. Includes a specific engine spec, most work done upstream on `elasticsearch-dbapi` (https://github.com/preset-io/elasticsearch-dbapi) that implements a PEP-249 DBAPI and SQLAlchemy dialect that "queries" X-Pack SQL (Limited to no support for opendistro yet)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Example Dashboard:

image

SQLLab:

image

### ADDITIONAL INFORMATION

- [X] Has associated issue: #5984 
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by villebro — APPROVED** [review]
Awesome! 🎉

**Review by mistercrunch — APPROVED** [review]
❤️

File: superset/db_engine_specs/elasticsearch.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        "P1Y": "HISTOGRAM({col}, INTERVAL 1 YEAR)",
+    }
+
+    type_code_map: Dict[int, str] = {}  # loaded from get_datatype only if needed
+
+    @classmethod
+    def convert_dttm(cls, target_type: str, dttm: datetime) -> str:
+        if target_type.upper() in ("DATETIME", "DATE"):
+            return f"'{dttm.isoformat()}'"
+        return "'{}'".format(dttm.strftime("%Y-%m-%d %H:%M:%S"))
```
Comment (john-bodley) [inline superset/db_engine_specs/elasticsearch.py]: Could we use an `f` string here?

Comment (dpgaspar) [reply superset/db_engine_specs/elasticsearch.py]: of course

**Review by john-bodley — APPROVED** [review]

**Comment by dioptre** [comment]
Wondering if anyone managed to get this working with opendistro? 

https://github.com/preset-io/elasticsearch-dbapi/issues/31

**Pull request merged**